File: java/01306_jump_game_III.java
// Time complexity - O (n), canReach is called for each index once
// Space complexity - O (n), dp and at most n recursive call

public class Solution {
    public boolean canReach(int[] arr, int start) {
        Boolean[] dp = new Boolean[arr.length];
        return canReach(start, arr, dp);
    }

    private boolean canReach(int position, int[] arr, Boolean[] dp) {
        if (position < 0 || position >= arr.length) {
            return false;
        }

        if (arr[position] == 0) {
            return true;
        }

        if (dp[position] != null) {
            return dp[position];
        }

        dp[position] = false;

        boolean canReachZero = canReach(position + arr[position], arr, dp) ||
                canReach(position - arr[position], arr, dp);

        dp[position] = canReachZero;
        return canReachZero;
    }
}

File: python/01306_jump_game_III.py
# Time complexity - O (n), canReach is called for each index once
# Space complexity - O (n), dp and at most n recursive call

class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        dp = [None] * len(arr)
        return self.canReachHelper(arr, start, dp)

    def canReachHelper(self, arr: List[int], position: int, dp: List[int]) -> bool:
        if position < 0 or position >= len(arr):
            return False

        if arr[position] == 0:
            return True

        if dp[position] is not None:
            return dp[position]

        dp[position] = False

        canReachZero = self.canReachHelper(arr, position + arr[position], dp) or \
                       self.canReachHelper(arr, position - arr[position], dp)

        dp[position] = canReachZero
        return dp[position]
